### PR TITLE
Common detector logic

### PR DESCRIPTION
**Pull request by JPrevost**
Note: this is the first of two change sets. The second will add a common singleton method in a separate module to aid in development of new or existing detectors. The second change set will wait for this to be approved and merged before being moved into review.

Why are these changes being introduced:

* We were starting to accumulate technical debt each time we added a
  new detector with a lot of similar (or duplicate) code

Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/TCO-104

How does this address that need:

* A new module was created `pattern_checker` and included into the
  detectors that could use it with minimal changes.

Document any side effects to this change:

* Two options where considered, and ultimately modules were chosen
* The other option was model inheritance which felt appropriate until
  considering that the parent model `Detector` has absolutely nothing
  in common with the individual detectors that are namespaced under it.
  Forcing this commonality was noticed when linters suggested that it
  is appropriate to call `super` when initializing a child class. We
  could of course skip that linter, and I considered it. But as I
  pondered whether I actually wanted to suggest that path, I tried the
  alternative path of extracting the common logic in a module, which
  doesn't have the side effect of potentially having inheritance for
  convenience rather than allowing `Detector` to only focus on what it
  is supposed to be managing for us.

### Accessibility

- [ ] ANDI or Wave has been run in accordance to [our guide](https://mitlibraries.github.io/guides/basics/a11y.html) and
      all issues introduced by these changes have been resolved or opened 
      as new issues (link to those issues in the Pull Request details above)
- [x] There are no accessibility implications to this change

### Documentation

- [x] Project documentation has been updated, and yard output previewed
- [ ] No documentation changes are needed

### ENV

- [ ] All new ENV is documented in README.
- [ ] All new ENV has been added to Heroku Pipeline, Staging and Prod.
- [x] ENV has not changed.

### Stakeholders

- [ ] Stakeholder approval has been confirmed
- [x] Stakeholder approval is not needed

### Dependencies and migrations

NO dependencies are updated

NO migrations are included

### Code

- [ ] I have confirmed that the code works as intended.
- [ ] Any CodeClimate issues have been fixed or confirmed as
added technical debt.

### Documentation

- [ ] The commit message is clear and follows our guidelines
      (not just this pull request message).
- [ ] The documentation has been updated or is unnecessary.
- [ ] New dependencies are appropriate or there were no changes.

### Testing

- [ ] There are appropriate tests covering any new functionality.
- [ ] No additional test coverage is required.

### CODE REVIEW

**Review by matt-bernhardt — APPROVED** [review]
I'm fine with this merging now, but I wonder whether the references to `term` I've flagged should now be `phrase` instead?

If that's coming in the second PR, then ignore me and this can merge as-is.

File: app/models/detector/lcsh.rb [hunk 1]
```diff
@@ -4,13 +4,16 @@ class Detector
...
-    attr_reader :identifiers
+    attr_reader :detections
+
+    # shared instance methods
+    include Detector::PatternChecker
 
     # For now the initialize method just needs to run the pattern checker. A space for future development would be to
     # write additional methods to look up the detected LCSH for more information, and to confirm that the phrase is
     # actually an LCSH.
     def initialize(term)
```
[the remark this answers is not in the document]
Comment (JPrevost) [reply app/models/detector/lcsh.rb:15]: I didn't change this after you comment earlier of wanting everything to be the same. I indeed want these to be phrases and not Terms but not all models are ready for that yet (Citation for instance that just landed today and was where I first brought up normalizing to Phrases lol).

I'll give this a look on Monday to see if it feels okay to normalize now and possibly setup a quick check in.

File: test/controllers/graphql_controller_test.rb [hunk 2]
```diff
@@ -87,7 +87,7 @@ class GraphqlControllerTest < ActionDispatch::IntegrationTest
     json = response.parsed_body
 
     assert_equal('nature', json['data']['logSearchEvent']['detectors']['journals'].first['title'])
-    assert_equal({"issns"=>["0028-0836", "1476-4687"]},
+    assert_equal({ 'issns' => %w[0028-0836 1476-4687] },
```
[the remark this answers is not in the document]
Comment (JPrevost) [reply test/controllers/graphql_controller_test.rb:90]: Yeah, I think I got all but one and the last one was going to require thinking so I left it.

File: app/models/detector/lcsh.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (JPrevost) [reply app/models/detector/lcsh.rb:15]: @matt-bernhardt I've opened https://mitlibraries.atlassian.net/browse/TCO-106 to rework all Detectors to use Strings and would like to address your feedback under that ticket. Could you review this PR with that in mind to see if we can land the refactor to shared logic and then I'll loop back on the name/pattern updates in the follow on ticket?

**Pull request merged**